classes/plugins/ThemePlugin.inc.php

```diff
@@ -28,107 +51,358 @@ function ThemePlugin() {
-		$request = $this->getRequest();
-		if ($this->getEnabled() && !defined('SESSION_DISABLE_INIT')) {
-			$templateManager = TemplateManager::getManager($request);
-			HookRegistry::register('PageHandler::displayCss', array($this, '_displayCssCallback'));
 
-			$context = $request->getContext();
+		if (defined('SESSION_DISABLE_INIT')) {
+			return false;
+		}
+
+		// Don't perform any futher operations if theme is not currently active
+		if (!$this->isActive()) {
+			return true;
+		}
+
+		// Fire an initialization method which themes should use to add
+		// styles, scripts and fonts
+		$this->init();
+
+		$this->_registerTemplates();
+		$this->_registerStyles();
+		$this->_registerScripts();
+
+		return true;
+	}
+
+	/**
+	 * The primary method themes should use to add styles, scripts and fonts,
+	 * or register hooks. This method is only fired for the currently active
+	 * theme.
+	 *
+	 * @return null
+	 */
+	public abstract function init();
+
+	/**
+	* Determine whether or not this plugin is currently active
+	*
+	* This only returns true if the theme is currently the selected theme
+	* in a given context. Use self::getEnabled() if you want to know if the
+	* theme is available for use on the site.
+	*
+	* @return boolean
+	*/
+	public function isActive() {
+		$request = $this->getRequest();
+		$context = $request->getContext();
+		if (is_a($context, 'Context')) {
+			$activeTheme = $context->getSetting('themePluginPath');
+		} else {
 			$site = $request->getSite();
-			$contextOrSite = $context?$context:$site;
+			$activeTheme = $site->getSetting('themePluginPath');
+		}
 
-			// Add the stylesheet.
-			if ($contextOrSite->getSetting('themePluginPath') == basename($path)) {
-				$dispatcher = $request->getDispatcher();
-				$templateManager->addStyleSheet($dispatcher->url($request, ROUTE_COMPONENT, null, 'page.PageHandler', 'css', null, array('name' => $this->getName())), STYLE_SEQUENCE_LATE);
+		return $activeTheme == basename($this->getPluginPath());
+	}
 
-				// If this theme uses templates, ensure they're given priority.
-				array_unshift(
-					$templateManager->template_dir,
-					$path = Core::getBaseDir() . DIRECTORY_SEPARATOR . $this->getPluginPath() . '/templates'
-				);
+	/**
+	 * Add a stylesheet to load with this theme
+	 *
+	 * Style paths with a .less extension will be compiled and redirected to
+	 * the compiled file.
+	 *
+	 * @param string $name A name for this stylesheet
+	 * @param string $path The path to this stylesheet, relative to the theme
+	 * @param array $args Optional arguments hash. Supported args:
+	 *   'context': Whether to load this on the `frontend` or `backend`.
+	 *      default: `frontend`
+	 *   'priority': Controls order in which styles are printed
+	 *   'addLess': Additional LESS files to process before compiling. Array
+	 */
+	public function addStyle($name, $path, $args = array()) {
+
+		// Pass a file path for LESS files
+		if (substr($path, -4) == 'less' ) {
```

Comment:
Not a big deal, but better IMO to use -5 and include the `.`. Could also...

    define(LESS_FILENAME_SUFFIX, '.less');
    ...
    if (substr($path, strlen(LESS_FILENAME_SUFFIX)) == LESS_FILENAME_SUFFIX) {

One less magic number.